safe/new_impact_function/impact_function.py

```diff
@@ -0,0 +1,829 @@
+# coding=utf-8
+"""InaSAFE Disaster risk assessment tool developed by AusAid -
+  *Impact Function.**
+
+Contact : ole.moller.nielsen@gmail.com
+
+.. note:: This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+"""
+
+from qgis.core import (
+    QgsMapLayer,
+    QgsCoordinateReferenceSystem,
+    QgsRectangle,
+    QgsVectorLayer,
+    QgsGeometry,
+    QgsFeature,
+    QgsField,
+)
+
+from PyQt4.QtCore import QVariant
+import logging
+
+from safe.definitions import post_processors
+from safe.defaults import get_defaults
+from safe.common.exceptions import InvalidExtentError, InvalidLayerError
+from safe.utilities.i18n import tr
+from safe.common.utilities import get_non_conflicting_attribute_name
+from safe.utilities.keyword_io import KeywordIO
+
+LOGGER = logging.getLogger('InaSAFE')
+
+
+__author__ = 'ismailsunni'
+__project_name__ = 'inasafe-dev'
+__filename__ = 'impact_function'
+__date__ = '8/16/16'
+__copyright__ = 'imajimatika@gmail.com'
+
+
+def evaluate_formula(formula, variables):
+    """Very simple formula evaluator. Beware the security.
+    :param formula: A simple formula.
+    :type formula: str
+
+    :param variables: A collection of variable (key and value).
+    :type variables: dict
+
+    :returns: The result of the formula execution.
+    :rtype: float, int
+    """
+    for key, value in variables.items():
+        formula = formula.replace(key, str(value))
+    return eval(formula)
+
+
+class ImpactFunction(object):
+    """Impact Function."""
+
+    def __init__(self):
+        self._hazard = None
+        self._hazard_keyword = {}
```

Comment:
We are going to use monkey patching to add  'keywords' to the 'QgsMapLayer'.
In the exposure, hazard and aggregation setters, you should monkey patch keywords.

The workflow will be to read keywords at the beginning of the IF and write them into the layer.keywords as a dictionary. Then in the impact function, we will only work with this monkey patching to insert/update/delete keywords. Only at the end of the impact function, we should use keyword_io to write them on disk.

So we should remove self._hazard_keyword etc

See PR https://github.com/inasafe/inasafe/pull/3219